internal/server/filesystem/server.go

```diff
@@ -2,24 +2,227 @@ package filesystem
+	// check for UNC/pipe prefixes like "\\"
+	if len(path) < 2 {
+		return false
+	}
+	if path[0] == '\\' && path[1] == '\\' {
+		return true
+	}
+	return false
+}
+
+func isAbsWindows(path string) bool {
+	// for Windows check for C:\\.. prefix only
+	// UNC prefixes of the form \\ are not considered
+	// absolute in the context of CSI proxy
+	if len(path) < 3 {
+		return false
+	}
+	c := path[0]
+	if !isDriveLetterWindows(c) {
+		return false
+	}
+	if path[1] != ':' {
+		return false
+	}
+	if path[2] != '\\' && path[2] != '/' {
+		return false
+	}
+	return true
+}
+
+func (s *Server) validatePathWindows(pathCtx internal.PathContext, path string) error {
+	prefix := ""
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+
+	pathlen := len(path)
+
+	if pathlen > MAX_PATH_LENGTH_WINDOWS {
+		return fmt.Errorf("Path length %d exceeds maximum characters: %d", pathlen, MAX_PATH_LENGTH_WINDOWS)
+	}
+
+	if len(path) > 0 && (path[0] == '\\') {
+		return fmt.Errorf("Invalid character \\ at begining of path: %s", path)
+	}
+
+	if isUNCPathWindows(path) {
+		return fmt.Errorf("Unsupported UNC path prefix: %s", path)
+	}
+
+	if containsInvalidCharactersWindows(path) {
+		return fmt.Errorf("Path contains invalid characters: %s", path)
+	}
+
+	if isAbsWindows(path) && !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("Absolute path: %s is not within context path: %s", path, prefix)
+	}
+
+	return nil
+}
+
+func (s *Server) abs(pathCtx internal.PathContext, path string) (string, error) {
+	if isAbsWindows(path) {
+		return path, nil
+	}
+	prefix := ""
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return "", fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+	return prefix + "\\" + path, nil
+}
 
 // PathExists checks if the given path exists on the host.
 func (s *Server) PathExists(ctx context.Context, request *internal.PathExistsRequest, version apiversion.Version) (*internal.PathExistsResponse, error) {
-	// FIXME: actually implement this!
+	err := s.validatePathWindows(request.Context, request.Path)
+	if err != nil {
+		return &internal.PathExistsResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	path, err := s.abs(request.Context, request.Path)
+	if err != nil {
+		return &internal.PathExistsResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	exists, err := s.hostAPI.PathExists(path)
 	return &internal.PathExistsResponse{
-		Success: false,
-		CmdletError: &api.CmdletError{
-			CmdletName: "dummy",
-			Code:       12,
-			Message:    "hey there " + request.Path,
-		},
+		Error:  err.Error(),
+		Exists: exists,
+	}, nil
+}
+
+func (s *Server) Mkdir(ctx context.Context, request *internal.MkdirRequest, version apiversion.Version) (*internal.MkdirResponse, error) {
```

Comment:
the passing parameter ctx context.Context is not used?

Reply:
Correct. `ctx` is not very relevant for quick operations like the FS ones and therefore not used. A rough parallel is `ctx` passed to all CSI API calls but drivers typically ignore it when servicing the API calls in the [Identity set](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/330efe6a568a25eb8430218ca568fe7f710b5622/pkg/gce-pd-csi-driver/identity.go#L30).